From f6b8ab35215edfc1cea7d05af3003069e094c133 Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Tue, 16 Mar 2021 18:45:40 +0200
Subject: [PATCH] [SW-0] gaudi2_arm: add support for spi image generation

Add support to generate SPI image for gaudi2 in legacy
mode using the latest scheme of image pointers.

Change-Id: Id0a59c53730ae8777e65c5065a268ebb387e6a81
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/board/habanalabs/gaudi2_arm/post-image-spi.sh b/board/habanalabs/gaudi2_arm/post-image-spi.sh
new file mode 100755
index 0000000..f175ac2
--- /dev/null
+++ b/board/habanalabs/gaudi2_arm/post-image-spi.sh
@@ -0,0 +1,39 @@
+#!/usr/bin/env bash
+
+function handle_btl {
+	local INF=$1
+	local OUTF=$2
+
+	for s in s?.bin; do
+		# Align to 4B
+		dd status=none conv=sync bs=4 if=${s} of=${s}.le
+	done
+
+	# Align to 64KB
+	dd status=none conv=sync bs=64K if=${INF} of=${INF}.tmp
+	swap_file ${INF}.tmp ${OUTF}
+}
+
+function create_spi_image {
+	local BTL_BIN=$1
+	local PREBOOT_BIN=$2
+	local IMG_PTR_BIN=$3
+	local OUTF=$4
+
+	# BTL
+	dd status=none conv=sync bs=64K		if=${BTL_BIN}		 > ${OUTF}
+	# Image pointers ARC and Security ARC
+	dd status=none conv=sync bs=128K count=1 if=128K-0xff.bin	>> ${OUTF}
+	# Image pointers ARM
+	dd status=none conv=sync bs=64K		if=${IMG_PTR_BIN}	>> ${OUTF}
+	# Security and U-Boot ENV
+	dd status=none conv=sync bs=128K count=1 if=128K-0xff.bin	>> ${OUTF}
+	# Preboot1
+	dd status=none conv=sync bs=64K	if=${PREBOOT_BINFILE}		>> ${OUTF}
+	# Preboot2
+	dd status=none conv=sync bs=64K	if=${PREBOOT_BINFILE}		>> ${OUTF}
+}
+
+PREBOOT_OFFSET=0xb0000
+PREBOOT2_OFFSET=0xc0000
+PREBOOT_TEXT_BASE=0x7FFD004000
diff --git a/board/habanalabs/gaudi2_arm/uboot-eeprom.cmd b/board/habanalabs/gaudi2_arm/uboot-eeprom.cmd
new file mode 100644
index 0000000..86759a2
--- /dev/null
+++ b/board/habanalabs/gaudi2_arm/uboot-eeprom.cmd
@@ -0,0 +1,19 @@
+#TODO: make changes related to zephyr agent
+setenv LAYOUT "-l legacy"
+mw.b 0x7ff0000000 ff 100
+eeprom write 0x7ff0000000 0 100
+eeprom update $LAYOUT "Layout Version" "0"
+eeprom update $LAYOUT "Production Serial Number" "a011223344"
+eeprom update $LAYOUT "PCB Assembly Version" "1a"
+eeprom update $LAYOUT "CPLD FW Version" "2b"
+eeprom update $LAYOUT "Infineon FW Version" "3c"
+eeprom update $LAYOUT "DDR Type and Speed" "12"
+eeprom update $LAYOUT "Fans Vendor" "3"
+eeprom update $LAYOUT "Mechanics Vendor" "4"
+eeprom update $LAYOUT "Thermal Solution Vendor" "5"
+eeprom update $LAYOUT "PCIe Gen3 or Gen4 Enabled" "6"
+eeprom update $LAYOUT "MAC Address" "00:11:22:33:44:55"
+eeprom update $LAYOUT "Board Configuration" "Gaudi2 HL1001"
+echo "eeprom content is:"
+eeprom print  $LAYOUT
+setenv LAYOUT
diff --git a/configs/gaudi2_arm_defconfig b/configs/gaudi2_arm_defconfig
index 9061ed3..50e0ee6 100644
--- a/configs/gaudi2_arm_defconfig
+++ b/configs/gaudi2_arm_defconfig
@@ -13,7 +13,7 @@
 #BR2_POST_IMAGE_UBOOT_ITS_TEMPLATE="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/uboot_extended.its"
 BR2_ROOTFS_OVERLAY="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/rootfs_overlay"
 BR2_ROOTFS_POST_BUILD_SCRIPT="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/post-build.sh"
-#BR2_ROOTFS_POST_IMAGE_SCRIPT="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/post-image-spi.sh"
+BR2_ROOTFS_POST_IMAGE_SCRIPT="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/post-image-spi.sh"
 BR2_LINUX_KERNEL=n
 BR2_LINUX_KERNEL_USE_ARCH_DEFAULT_CONFIG=y
 BR2_LINUX_KERNEL_CONFIG_FRAGMENT_FILES="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/$(HL_DEV)/linux_$(HL_TARGET).fragment"
